src/libsyntax/ext/derive.rs

```diff
@@ -0,0 +1,225 @@
+
+        // First, weed out malformed #[derive]
+        traits.retain(|titem| titem.word().is_some());
+
+        let mut titem = None;
+
+        // See if we can find a matching trait.
+        for j in 0..traits.len() {
+            if derive_type.is_derive_type(cx, &traits[j]) {
+                titem = Some(traits.remove(j));
+                break;
+            }
+        }
+
+        // If we find a trait, remove the trait from the attribute.
+        if let Some(titem) = titem {
+            if traits.len() == 0 {
+                attrs.remove(i);
+            } else {
+                let derive = Symbol::intern("derive");
+                let mitem = cx.meta_list(titem.span, derive, traits);
+                attrs[i] = cx.attribute(titem.span, mitem);
+            }
+            let derive = Symbol::intern("derive");
+            let mitem = cx.meta_list(titem.span, derive, vec![titem]);
+            return Some(cx.attribute(mitem.span, mitem));
+        }
+    }
+    return None;
+}
+
+pub fn get_legacy_derive(cx: &mut ExtCtxt, attrs: &mut Vec<ast::Attribute>)
+                         -> Option<ast::Attribute> {
+    get_derive_attr(cx, attrs, DeriveType::Legacy).and_then(|a| {
+        let titem = derive_attr_trait(cx, &a);
+        if let Some(titem) = titem {
+            let tword = titem.word().unwrap();
+            let tname = tword.name();
+            if !cx.ecfg.enable_custom_derive() {
+                feature_gate::emit_feature_err(&cx.parse_sess,
+                                               "custom_derive",
+                                               titem.span,
+                                               feature_gate::GateIssue::Language,
+                                               feature_gate::EXPLAIN_CUSTOM_DERIVE);
+            } else {
+                let name = Symbol::intern(&format!("derive_{}", tname));
+                if !cx.resolver.is_whitelisted_legacy_custom_derive(name) {
+                    cx.span_warn(titem.span, feature_gate::EXPLAIN_DEPR_CUSTOM_DERIVE);
+                }
+                let mitem = cx.meta_word(titem.span, name);
+                return Some(cx.attribute(mitem.span, mitem));
+            }
+        }
+        None
+    })
+}
+
+pub fn get_builtin_derive(cx: &mut ExtCtxt, attrs: &mut Vec<ast::Attribute>)
+                          -> Option<ast::Attribute> {
+    get_derive_attr(cx, attrs, DeriveType::Builtin)
+}
+
+pub fn get_proc_macro_derive(cx: &mut ExtCtxt, attrs: &mut Vec<ast::Attribute>)
+                             -> Option<ast::Attribute> {
+    get_derive_attr(cx, attrs, DeriveType::ProcMacro)
+}
+
+pub fn is_legacy_derive(cx: &mut ExtCtxt, titem: &NestedMetaItem) -> bool {
+    let tname = titem.name().unwrap();
+    let name = Symbol::intern(&format!("derive_{}", tname));
+    let path = ast::Path::from_ident(titem.span, ast::Ident::with_empty_ctxt(name));
+        cx.resolver.resolve_macro(cx.current_expansion.mark, &path, false).is_ok()
```

Comment:
Ideally we would just resolve in `resolver.builtin_macros`, either by moving this logic into `resolve` or adding a method to check builtin macros directly to `ext::base::Resolver`.